packages/core-api/src/routing/hooks.tsx

```diff
@@ -0,0 +1,183 @@
+      currentRouteRef = this.routeParents.get(currentRouteRef)
+    ) {
+      matchIndex = match.findIndex(
+        m => (m.route as BackstageRouteObject).routeRef === currentRouteRef,
+      );
+      if (matchIndex !== -1) {
+        break;
+      }
+
+      targetRefStack.unshift(currentRouteRef);
+    }
+
+    // If our target route is present in the initial match we need to construct the final path
+    // from the parent of the matched route segment. That's to allow the caller of the route
+    // function to supply their own params.
+    if (targetRefStack.length === 0) {
+      matchIndex -= 1;
+    }
+
+    // This is the part of the route tree that the target and source locations have in common.
+    // We re-use the existing pathname directly along with all params.
+    const parentPath = matchIndex === -1 ? '' : match[matchIndex].pathname;
+
+    // This constructs the mid section of the path using paths resolved from all route refs
+    // we need to traverse to reach our target except for the very last one. None of these
+    // paths are allowed to require any parameters, as the called would have no way of knowing
+    // what parameters those are.
+    const prefixPath = targetRefStack
+      .slice(0, -1)
+      .map(ref => {
+        const path = this.routePaths.get(ref);
+        if (!path) {
+          throw new Error(`No path for ${ref}`);
+        }
+        if (path.includes(':')) {
+          throw new Error(
+            `Cannot route to ${routeRef} with parent ${ref} as it has parameters`,
+          );
+        }
+        return path;
+      })
+      .join('/')
+      .replace(/\/\/+/g, '/'); // Normalize path to not contain repeated /'s
+
+    const routeFunc: RouteFunc<Params> = (...[params]) => {
+      return `${parentPath}${prefixPath}${generatePath(lastPath, params)}`;
+    };
+    return routeFunc;
+  }
+}
+
+const RoutingContext = createContext<RouteResolver | undefined>(undefined);
+
+export function useRouteRef<Params extends { [param in string]: string }>(
+  routeRef: RouteRef<Params>,
+): RouteFunc<Params> {
+  const sourceLocation = useLocation();
+  const resolver = useContext(RoutingContext);
+  const routeFunc = useMemo(
+    () => resolver && resolver.resolve(routeRef, sourceLocation),
+    [resolver, routeRef, sourceLocation],
+  );
+
+  if (!routeFunc) {
+    throw new Error('No route resolver found in context');
+  }
+
+  return routeFunc;
+}
+
+type ProviderProps = {
+  routePaths: Map<AnyRouteRef, string>;
+  routeParents: Map<AnyRouteRef, AnyRouteRef | undefined>;
+  routeObjects: BackstageRouteObject[];
+  children: ReactNode;
+};
+
+export const RoutingProvider = ({
+  routePaths,
+  routeParents,
+  routeObjects,
+  children,
+}: ProviderProps) => {
+  const resolver = new RouteResolver(routePaths, routeParents, routeObjects);
+  return (
+    <RoutingContext.Provider value={resolver}>
+      {children}
+    </RoutingContext.Provider>
+  );
+};
+
+export function validateRoutes(
```

Comment:
I wonder if it's worth documenting this function and what it does so far too right?

Reply:
Yep, this is probably never going to be public though so perhaps we can do it in a followup